geo/src/algorithm/haversine_intermediate.rs

```diff
@@ -0,0 +1,118 @@
+use num_traits::{Float, FromPrimitive};
+use ::Point;
+
+/// Returns a new Point along a great circle route between two existing points
+
+pub trait HaversineIntermediate<T: Float> {
+    /// Returns a new Point along a great circle route between two existing points.
+    ///
+    /// ```
+    /// # extern crate geo;
+    /// # #[macro_use] extern crate approx;
+    /// #
+    /// use geo::Point;
+    /// use geo::algorithm::haversine_intermediate::HaversineIntermediate;
+    ///
+    /// # fn main() {
+    /// let p1 = Point::<f64>::new(10.0,  20.0);
+    /// let p2 = Point::<f64>::new(125.0, 25.0);
+    /// let i20        = p1.haversine_intermediate(&p2, 0.2);
+    /// let i50        = p1.haversine_intermediate(&p2, 0.5);
+    /// let i80        = p1.haversine_intermediate(&p2, 0.8);
+    /// let i20_should = Point::new(29.8,  29.9);
+    /// let i50_should = Point::new(65.8,  37.7);
+    /// let i80_should = Point::new(103.5, 33.5);
+    /// assert_relative_eq!(i20.x(), i20_should.x(), epsilon=0.2);
+    /// assert_relative_eq!(i20.y(), i20_should.y(), epsilon=0.2);
+    /// assert_relative_eq!(i50.x(), i50_should.x(), epsilon=0.2);
+    /// assert_relative_eq!(i50.y(), i50_should.y(), epsilon=0.2);
+    /// assert_relative_eq!(i80.x(), i80_should.x(), epsilon=0.2);
+    /// assert_relative_eq!(i80.y(), i80_should.y(), epsilon=0.2);
+    /// # }
+    /// ```
+    fn haversine_intermediate(&self, other: &Point<T>, f: T) -> Point<T>;
+}
+
+impl<T> HaversineIntermediate<T> for Point<T>
+where
+    T: Float + FromPrimitive,
+{
+    fn haversine_intermediate(&self, other: &Point<T>, f: T) -> Point<T> {
+        let one = T::one();
+        let two = one + one;
+
+        let lat1 =  self.y().to_radians();
+        let lon1 =  self.x().to_radians();
+        let lat2 = other.y().to_radians();
+        let lon2 = other.x().to_radians();
+
+        let k = (
+            ((lat1 - lat2) / two).sin().powi(2) +
+            lat1.cos() * lat2.cos() * ((lon1 - lon2) / two).sin().powi(2)
+        ).sqrt();
+
+        let d = two * k.asin();
+
+        let a = ((one - f) * d).sin() / d.sin();
+        let b = (f * d).sin() / d.sin();
+
+        let x = a * lat1.cos() * lon1.cos() + b * lat2.cos() * lon2.cos();
+        let y = a * lat1.cos() * lon1.sin() + b * lat2.cos() * lon2.sin();
+        let z = a * lat1.sin() + b * lat2.sin();
+
+        let dxy = (x.powi(2) + y.powi(2)).sqrt();
```

Comment:
I think this should be the same as `x.hypot(y)`

https://docs.rs/num/0.1.42/num/trait.Float.html#tymethod.hypot

Reply:
agreed!